// cSettings/csettings.js
var cSettings = {};

function init(path_to_config, callback) {
	var json = '';
	var path = path_to_config || process.cwd() + "/config.json";
	if (path.substring(0,7) === 'http://')
	{
		var url = require('url').parse(path, true, true);
		var options={
			host: url.host,
			port: 80,
			path: url.pathname
		};
		require("http").get(options, function(get_res) {
			get_res.on('data', function(chunk) {
				json += chunk.toString();
			});
			get_res.on('end', function() {
				parseSettings(JSON.parse(json), callback);
			});
		}).on('error', function(err) {
			console.error("ERROR: cSettings: could not get settings file: " + JSON.stringify(err));
		});

	}
	else if (path.substring(0,8) === 'mongo://'){
		var db = require('mongoskin').db(path);
		db.collection('settings').findOne({}, function(err, config){
			if (err) throw err;
			parseSettings(config, callback);
		});
	}
	else {
		json = JSON.parse(require('fs').readFileSync(path, 'utf8'));
		parseSettings(json, callback);
	}
}

function parseSettings(json, callback){
	//handle port on hosting portals
	json.env.port = process.env['app_port'] || json.env.port;
	// stringify db url
	//for(var idb in json.db){
	//	var db = json.db[idb];
	//	db.url =  db.driver + '://' + db.user + ':' + db.pw;
	//	db.url += '@' + db.host + ':' + db.port + '/' + db.db;
	//	if (db.driver == 'mongo'){
	//		db.url += '?auto_reconnect';
	//	}
	//}
	module.exports = json;
	callback(json);
}

module.exports.init = init;
