// cOS/cOS.js
var os = require('os');
var common = require('cCommon');

exports.route = [{
   api: 'os',
   method: 'GET',
   callback: function(req, res){
     var cpu = os.cpus();
     var s = {
       'Hostname' : os.hostname(),
       'OS name' : os.type(),
       //s.platform = os.platform();
       //s.arch = os.arch();
       'Release' : os.release(),
       'Uptime'  : common.secondsToDHms(os.uptime()),
       'Load' : roundLoadAVG(os.loadavg()),
       'Total Mem': common.bytesToSize(os.totalmem(), 3),
       'Free Mem' : common.bytesToSize(os.freemem(), 3),
       'CPUs'    : cpu.count,
       'CPU Load' : calcCPULoad(cpu)
       //'networkinterfaces' : os.networkInterfaces()

      };
      res.end(JSON.stringify(s));
   }
}];

function calcCPULoad(cpus){
    var ticks = 0;
    var idle = 0;
    cpus.forEach(function(cpu){
       ticks += cpu.times.sys + cpu.times.user + cpu.times.nice + cpu.times.idle + cpu.times.irq; 
       idle += cpu.times.idle;
    });
    return ((ticks - idle) / ticks).toFixed(4) * 100 + ' %';
}

function roundLoadAVG(avgs){
    var s = '';
    avgs.forEach(function(avg){
        s += avg.toFixed(3) + ', '; 
    });
    return s.substr(0,s.length -2);
}
