// cDB/cdb.js
var cSettings = require('cSettings');
var db = require('mongoskin').db(cSettings.db);

exports.route = [{
   api: 'db',
   method: 'GET',
   callback: get
}, {
   api: 'db',
   method: 'POST',
   callback: post
}, {
   api: 'db',
   method: 'PUT',
   callback: put
}, {
   api: 'db',
   method: 'DELETE',
   callback: del
}];

function get(req, res, callback) {
   parse(req, res, _get, callback);
}

function post(req, res, callback) {
   parse(req, res, _post, callback);
}

function put(req, res, callback) {
   parse(req, res, _put, callback);
}

function del(req, res, callback) {
   parse(req, res, _del, callback);
}


function parse(req, res, action, callback) {
   var collection = req.cURL.path[0];
   if (collection === 'users') {
      res.end("Not allowed"); // TODO: route to 404 page
      return;
   }
   if (!callback){
      callback = function(result){
         var parsedresult = JSON.stringify(result);
         res.end(parsedresult);
      };
   }
   action(req, res, collection, callback);
}

function _get(req, res, coll, callback) {
   db.collection(coll).find({
      "user": req.user.id + ""
   }).toArray(function (err, elements) {
      if (err) callback(err);
      else callback(elements);
   });
}

function _post(req, res, coll, callback) {
   req.post.user = req.user.id;
   db.collection(coll).insert(req.post, function (err, replies) {
      if (err) callback(err);
      else callback('OK');
   });
}

function _put(req, res, coll, callback) {}

function _del(req, res, coll, callback) {}

exports.get = get;
exports.post = post;
exports.put = put;
exports.del = del;
